sphinxgallery/gen_rst.py

```diff
@@ -118,133 +79,134 @@ def flush(self):
+
+    Returns
+    -------
+    blocks : list of (label, content)
+        List where each element is a tuple with the label ('text' or 'code'),
+        and content string of block.
     """
-    lines = open(filename).readlines()
-    start_row = 0
-    if lines[0].startswith('#!'):
-        lines.pop(0)
-        start_row = 1
-    docstring = ''
-    first_par = ''
-    line_iterator = iter(lines)
-    tokens = tokenize.generate_tokens(lambda: next(line_iterator))
-    for tok_type, tok_content, _, (erow, _), _ in tokens:
-        tok_type = token.tok_name[tok_type]
-        if tok_type in ('NEWLINE', 'COMMENT', 'NL', 'INDENT', 'DEDENT'):
-            continue
-        elif tok_type == 'STRING':
-            docstring = eval(tok_content)
-            # If the docstring is formatted with several paragraphs, extract
-            # the first one:
-            paragraphs = '\n'.join(
-                line.rstrip() for line
-                in docstring.split('\n')).split('\n\n')
-            if paragraphs:
-                if ignore_heading:
-                    if len(paragraphs) > 1:
-                        first_par = re.sub('\n', ' ', paragraphs[1])
-                        first_par = ((first_par[:95] + '...')
-                                     if len(first_par) > 95 else first_par)
-                    else:
-                        raise ValueError("Docstring not found by gallery.\n"
-                                         "Please check the layout of your"
-                                         " example file:\n {}\n and make sure"
-                                         " it's correct".format(filename))
-                else:
-                    first_par = paragraphs[0]
-
-        break
-    return docstring, first_par, erow + 1 + start_row
-
-
-def extract_line_count(filename, target_dir):
-    """Extract the line count of a file"""
-    example_file = os.path.join(target_dir, filename)
-    lines = open(example_file).readlines()
-    start_row = 0
-    if lines and lines[0].startswith('#!'):
-        lines.pop(0)
-        start_row = 1
-    line_iterator = iter(lines)
-    tokens = tokenize.generate_tokens(lambda: next(line_iterator))
-    check_docstring = True
-    erow_docstring = 0
-    for tok_type, _, _, (erow, _), _ in tokens:
-        tok_type = token.tok_name[tok_type]
-        if tok_type in ('NEWLINE', 'COMMENT', 'NL', 'INDENT', 'DEDENT'):
+
+    with open(source_file) as f:
+        source_lines = f.read()
+
+    doc_string = ast.get_docstring(ast.parse(source_lines))
+
+    blocks = [('text', doc_string)]
+    # Remove from source_lines the docstring
+    end_doc = [m.end() for m in re.finditer('"""', source_lines)][1]
+    source_lines = source_lines[end_doc:]
+
+    pattern = re.compile('^#{20,}.*\s((?:^#.*\s)*)', flags=re.M)
+    split_block = re.split(pattern, source_lines)
+    for split_code_block in split_block:
+        if split_code_block is None or split_code_block.strip() == '':
```

Comment:
I tried a few things but I couldn't figure out a way to get a None for split_code_block. Do you have one in mind ?

I would put a comment on top of this condition to make it more clear when it happens (e.g. newlines between two comment blocks).

Additional bonus points if you improve the naming for `split_block` and `split_code_blocks`. Maybe `for content in block_contents`? More than open for better suggestions.